candle-metal-kernels/src/lib.rs

```diff
@@ -270,31 +270,32 @@ impl Kernels {
 #[allow(clippy::too_many_arguments)]
 pub fn call_unary_contiguous(
     device: &Device,
-    command_buffer: &CommandBufferRef,
+    encoder: &ComputeCommandEncoderRef,
     kernels: &Kernels,
     kernel_name: unary::contiguous::Kernel,
     length: usize,
     input: BufferOffset,
     output: &Buffer,
 ) -> Result<(), MetalKernelError> {
     let pipeline = kernels.load_pipeline(device, Source::Unary, kernel_name.0)?;
-    let encoder = command_buffer.new_compute_command_encoder();
     encoder.set_compute_pipeline_state(&pipeline);
 
     set_params!(encoder, (length, &input, output));
 
     let (thread_group_count, thread_group_size) = linear_split(&pipeline, length);
+
+    encoder.memory_barrier_with_resources(&[input.buffer]);
```

Comment:
Isn't this enforced by the `use_resource` just below? I'm not 100% sure of the semantics of these two so any link to some doc would be great!

Reply:
Good question, in this case useResource is more about granting visibility of the memory, it doesn't gate execution of the kernel on the resource. The memory barrier is used to ensure that operations involving the input buffers have completed and that we can safely read from those buffers.

Reply:
I think there is a lot of room for improvement in reducing these calls so that we can avoid the candle_metal_kernels code from needing to manage access requests, but that would involve knowing the compute graph prior to any synchronization event.

Reply:
Not sure to understand what you mean here. My mental model based on cuda is that we inform metal of the dependencies and it will keep track of the compute graph internally. The only place where you need synchronisation is when sending or retrieving data or when doing very weird things in kernels (which we don't in candle). Is that actually different in metal?

Reply:
That's true across encoders/buffers, but when we are executing many kernels on a single encoder metal does not provide that guarantee. This is at least my understanding, I'll play around with different configurations of the useResources and memoryBarrierWithResources to conrfirm though.

Reply:
So it seems that the following is the behaviour under different scenarios:
- no odd behaviour when using both useResources & memoryBarrierWithResources
- no odd behaviour when only using memoryBarrierWithResources
- no odd behaviour when only using useResources
- no odd behaviour when using neither

Now this is odd, I'm not entirely convinced from a quick pass at the examples that we'd be getting this correct though... looking back at llama.cpp, they seem to only leverage the memorybarrier op and not the useresource op. I'm inclined to believe that we're goth correct here and that I suppose we just need one or the other?

Reply:
@ivarflakstad do you have any insight here that could help us land on the preferred option?

Reply:
> That's true across encoders/buffers, but when we are executing many kernels on a single encoder metal does not provide that guarantee. This is at least my understanding

Would you have any link for this?
I certainly would rather see the specs rather than trying it out as getting it wrong would make for bugs that are very tricky to debug.
Here is the most relevant bit I found.
> Compute commands are executed in the order in which they are encoded into the command buffer. A compute command finishes execution when all threadgroups associated with the command finish execution and all results are written to memory. Because of this sequencing, the results of a compute command are available to any commands encoded after it in the command buffer.
It's coming from this [doc](https://developer.apple.com/library/archive/documentation/Miscellaneous/Conceptual/MetalProgrammingGuide/Compute-Ctx/Compute-Ctx.html#//apple_ref/doc/uid/TP40014221-CH6-SW1), and I understand that as long as you're using the same buffer you're find so you would need barries when you have multiple buffers in parallel?

Reply:
I've mostly been using these docs for informing myself https://developer.apple.com/documentation/metal/resource_synchronization?language=objc

Re-reading, perhaps I'm understanding wrong and these are automatically tracked for us since we are creating buffers from the device interface, and the hazard tracking is handled for us

Reply:
However this doc, which I believe is the more relevant one for us seems to indicate the useResource is required for the argument buffer bindings we make https://developer.apple.com/documentation/metal/buffers/tracking_the_resource_residency_of_argument_buffers?language=objc